# <span style="color:blue"> 이토록 쉬운 통계 & R </span>

<img src="../image/cover.png" width="200" height="400" align="left">  임경덕 지음

## PART 4 데이터 분석 도구, R

## 90% 순서대로 살펴보는 BR31-3 (33일차)

   * 데이타프로젝트 (http://dataproject.xyz) 또는 (http://cafe.naver.com/dataartproject) 에 미리 저장된 스크립트를 다운 받을 수 있습니다.
   * 스크립트를 활용하면 직접 명령어를 입력하지 않고 저장된 명령어들을 실행할 수 있습니다.
   * 지금 설치된 R 버젼은 3.4.3 입니다. 제목 BR31 은 Basic R 31 로 3.1 버젼입니다. 그러나 큰 차이는 없을 것 같습니다.
   * page 262 read.csv() 함수부터


### 데이터 파일은 dataartproject.xyz 에 있습니다.

In [3]:
# read.csv 함수로 csv 파일 읽어오기
# csv 는 comma seperated value 로 필드가 구분을 comma 로 합니다.
# 현재 폴더에 Practive/data 폴더 안에 파일들이 들어 있습니다.
read.csv("Practice/data/sample1.csv")

Gender Salary
1 M      100   
2 M      200   
3 F      300   
4 M      400   
5 F      500

In [5]:
# sample1.csv 파일을 읽어 data1 에 넣습니다.
data1 = read.csv("Practice/data/sample1.csv")

In [6]:
# data1 표시
data1

Gender Salary
1 M      100   
2 M      200   
3 F      300   
4 M      400   
5 F      500

In [11]:
head(data1)

Gender Salary
1 M      100   
2 M      200   
3 F      300   
4 M      400   
5 F      500

   * 5 개의 관측치가 있습니다.
   * 2개의 변수가 있으면 각각 변수명은 Gender, Salary 입니다.
   * 파일을 열어보면, header(컬렴명) 가 있습니다.

In [7]:
# sample2 파일을 보면
read.csv("Practice/data/sample2.csv")

High   X8
1 Low    6 
2 Middle 7

   * Sample 2 인 경우에는 header 가 없습니다.
   * 이 경우에는 header option 을 설정합니다.

In [9]:
# header option 을 false 로 하면, 자동적으로 변수명을 v1, v2,... 로 지정합니다.
read.csv("Practice/data/sample2.csv", header=FALSE)

V1     V2
1 High   8 
2 Low    6 
3 Middle 7

In [46]:
# head() 함수로 관측치 확인하기
subway = read.csv("Practice/data/subway.csv")

In [47]:
head(subway)

역명        날짜       구분 X05.06시 X06.07시 X07.08시 X08.09시 X09.10시
1 서울역(150) 2016-12-01 승차 411       552     1792      3429    2737    
2 서울역(150) 2016-12-01 하차 280      1680     4629     10313    5586    
3 서울역(150) 2016-12-02 승차 434       520     1695      3456    2848    
4 서울역(150) 2016-12-02 하차 255      1609     4504     10225    5322    
5 서울역(150) 2016-12-03 승차 543       467      982      2080    3000    
6 서울역(150) 2016-12-03 하차 260      1355     1853      2337    3223    
  X10.11시 X11.12시 ⋯ X16.17시 X17.18시 X18.19시 X19.20시 X20.21시 X21.22시
1 2430     3161     ⋯ 4145     4984      9595    5217     3492     3978    
2 2920     2673     ⋯ 2905     3084      3715    2723     1794     1657    
3 2568     3398     ⋯ 5180     6141     10682    6169     4169     4479    
4 3050     2435     ⋯ 3692     3545      5575    3389     2141     2076    
5 3077     4340     ⋯ 8023     6418      6575    4904     4742     4683    
6 3218     3230     ⋯ 5384     6014      5494    3803     3440     3771    
  X22.23시 X23.24시 X00.01시 X01.02시
1 2742     1380     198        0     
2 1533      761     373       32     
3 4554     1794     443        0     
4 1928     1074     547       64     
5 4654     1697     228        0     
6 3317     1642     837      131

   * subway.csv 파일은 cp-949 한글코드로 되어 있으므로, 일단 utf-8 으로 변경합니다. 윈도우에서는 문제 없으나, 맥이나 리눅스에서는 문제가 있음.
   * 총 7,378 개의 관측치가 있으며, head() 함수에는 그중 앞에 6개만 출력합니다.
   * 처음 변수는 역명, 두번째 변수는 날짜, 세번째 변수는 승차 하차 구분, 이후에는 1시간 단위로 21개의 시간대 입니다.
   * X05.06 은 원본에서는 $05 \sim 06$ 으로 되어 있습니다.

   * 첫번째 관측치는 $5 \sim 6$ 에 승차한 사람수는 411명, $16 \sim 17$ 에는 4145 명 승차했습니다.

**names()** 함수로 변수 이름 확인하고 바꾸기

   * 데이터 분석은 변수 단위.
       * 예를 들어 키 변수를 가져와서 m 단위로 바꾸고 평균을 구하자!  이런식입니다.
   * R 에서는 데이터의 변수들이 어떤 이름으로 저장 되어 있는 지 확인 필요
   * 변수이름은 가급적 영어이름으로 변경

In [48]:
names(subway)

[1] "역명"     "날짜"     "구분"     "X05.06시" "X06.07시" "X07.08시"
 [7] "X08.09시" "X09.10시" "X10.11시" "X11.12시" "X12.13시" "X13.14시"
[13] "X14.15시" "X15.16시" "X16.17시" "X17.18시" "X18.19시" "X19.20시"
[19] "X20.21시" "X21.22시" "X22.23시" "X23.24시" "X00.01시" "X01.02시"

In [49]:
# 변수명 조정
names(subway) = c("Station", "Date", "InOut", paste0("H", 5:25))

In [50]:
names(subway)

[1] "Station" "Date"    "InOut"   "H5"      "H6"      "H7"      "H8"     
 [8] "H9"      "H10"     "H11"     "H12"     "H13"     "H14"     "H15"    
[15] "H16"     "H17"     "H18"     "H19"     "H20"     "H21"     "H22"    
[22] "H23"     "H24"     "H25"

In [51]:
# str() 함수로 데이터 구조 확인하기
str(subway)

'data.frame':	7378 obs. of  24 variables:
 $ Station: Factor w/ 119 levels "가락시장(340)",..: 52 52 52 52 52 52 52 52 52 52 ...
 $ Date   : Factor w/ 31 levels "2016-12-01","2016-12-02",..: 1 1 2 2 3 3 4 4 5 5 ...
 $ InOut  : Factor w/ 2 levels "승차","하차": 1 2 1 2 1 2 1 2 1 2 ...
 $ H5     : num  411 280 434 255 543 260 372 179 499 465 ...
 $ H6     : num  552 1680 520 1609 467 ...
 $ H7     : num  1792 4629 1695 4504 982 ...
 $ H8     : num  3429 10313 3456 10225 2080 ...
 $ H9     : num  2737 5586 2848 5322 3000 ...
 $ H10    : num  2430 2920 2568 3050 3077 ...
 $ H11    : num  3161 2673 3398 2435 4340 ...
 $ H12    : num  3433 2506 3926 3037 5121 ...
 $ H13    : num  3364 2393 3405 3103 4781 ...
 $ H14    : num  2767 2648 2876 2888 4335 ...
 $ H15    : num  3822 2535 5175 3391 6565 ...
 $ H16    : num  4145 2905 5180 3692 8023 ...
 $ H17    : num  4984 3084 6141 3545 6418 ...
 $ H18    : num  9595 3715 10682 5575 6575 ...
 $ H19    : num  5217 2723 6169 3389 4904 ...
 $ H20    : num  34

   * 데이터 형식은 data.frame 입니다.
   * 관측치 갯수는 7378 obs 이고 변수는 24개 입니다.
   * 다음으로 24개 변수의 특징이 있습니다.
   * Station, Date, InOut 은 범주형 변수이며, 각각 119 수준, 31 수준, 2 수준으로 되어 있습니다.

In [54]:
# view() 함수로 스프레드시트 확인 하기  -- View is not support in Jupyter R kernel
View(subway)

ERROR: Error in View(subway): ‘View()’ not yet supported in the Jupyter R kernel
